From dadb3cf4a80024629f9175f2a36b463198266290 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Fri, 26 Jan 2024 18:48:38 +0100
Subject: [PATCH 0562/1695] drivers: stm32_rifsc: add RISUP logic
 reconfiguration API

Adds stm32_rifsc_reconfigure_risup() API that allows a RISUP
reconfiguration at run time.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 7b0e5913aa2e142c3fbf83dfc877a3622ae60bf3 drivers: stm32_rifsc: add RISUP logic reconfiguration API]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: I3f971627f667b884e254ffd6f22d1c4f5c1c5775
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/354978
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 39 +++++++++++++++++++++++++++++
 core/include/drivers/stm32_rif.h    | 19 ++++++++++++++
 2 files changed, 58 insertions(+)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 46e395122..9b506f71c 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -515,6 +515,45 @@ static TEE_Result stm32_rimu_setup(struct rifsc_platdata *pdata)
 	return 0;
 }
 
+TEE_Result stm32_rifsc_reconfigure_risup(unsigned int risup_id,
+					 unsigned int cid,
+					 bool sec, bool priv, bool cfen)
+{
+	unsigned int offset = sizeof(uint32_t) *
+			      (risup_id / _PERIPH_IDS_PER_REG);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct risup_cfg *risup = NULL;
+
+	if (risup_id > rifsc_pdata.drv_data->nb_risup ||
+	    cid > MAX_CID_SUPPORTED)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	risup = &rifsc_pdata.risup[risup_id];
+
+	if (io_read32(rifsc_pdata.base + _RIFSC_RISC_RCFGLOCKR0 + offset) &
+	    BIT(risup_id % _PERIPH_IDS_PER_REG)) {
+		DMSG("RIMU configuration is locked");
+		return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	risup->cid_attr = cid << RIFSC_RISC_SCID_SHIFT;
+	if (cfen)
+		risup->cid_attr |= RIFSC_RISC_CFEN_MASK;
+	else
+		risup->cid_attr &= ~RIFSC_RISC_CFEN_MASK;
+
+	risup->sec = sec;
+	risup->priv = priv;
+
+	res = stm32_risup_cfg(&rifsc_pdata, risup);
+	if (res) {
+		EMSG("RISUP %u reconfiguration error", risup_id);
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
 TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
 					unsigned int master_cid,
 					bool cid_sel, bool sec, bool priv)
diff --git a/core/include/drivers/stm32_rif.h b/core/include/drivers/stm32_rif.h
index 88348d2dc..e7768a85e 100644
--- a/core/include/drivers/stm32_rif.h
+++ b/core/include/drivers/stm32_rif.h
@@ -97,6 +97,25 @@ static inline TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
 }
 #endif
 
+/*
+ * Reconfigures a specific RISUP
+ *
+ * @risup_id:	ID of the RISUP to reconfigure.
+ * @cid:	CID to set.
+ * @sec:	Peripheral is set to secure/non-secure
+ * @priv:	Peripheral is set to privileged/non-privileged
+ * @cfen:	Activate cid filetring
+ *
+ * Returns TEE_SUCCESS in case of success.
+ * Returns TEE_ERROR_BAD_PARAMETERS if either the RISUP ID or the master cid is
+ * incorrect.
+ * Returns TEE_ERROR_ACCESS_DENIED if the RISUP can't be reconfigured.
+ * Returns any TEE_Result compliant code in case of error.
+ */
+TEE_Result stm32_rifsc_reconfigure_risup(unsigned int risup_id,
+					 unsigned int cid,
+					 bool sec, bool priv, bool cfen);
+
 /*
  * Reconfigures a specific RIMU
  *
-- 
2.47.2

